### PR TITLE
PBR shader code refactoring

### PR DESCRIPTION
**Pull request by Popov72**
We have been discussing this with @sebavan.

The goal is to make the PBR shader code easier to read/maintain and to use when creating the new NME blocks.

I hope I didn't break anything, the unit tests are all ok and the new test coverage too...

Let me know what you think and discuss about those changes.

### CODE REVIEW

**Comment by Popov72** [comment]
How to run the tests locally in WebGL 1?

File: src/Shaders/ShadersInclude/pbrBlockAlbedoOpacity.fx [hunk 1]
```diff
@@ -0,0 +1,73 @@
+struct albedoOpacityOutParams
```
[the remark this answers is not in the document]
Comment (sebavan) [reply src/Shaders/ShadersInclude/pbrBlockAlbedoOpacity.fx]: Ok I love the consistency but I really guess we should analyse the perf impact here.

**Comment by sebavan** [comment]
The first positive result is that the test server does not look like it is taking longer to run through all the tests which is already a great sign.

**Comment by Popov72** [comment]
> Looks overall good to me I am just wondering how we could measure the perf impacts ?
> 
> Maybe you could try to create a scene where you would be almost all bound on gpu and double check the frame rate to be sure ???

Ok, going to do that as soon as possible.

**Comment by Popov72** [comment]
Tested on my GTX 1080, with vsync off in Chrome by using those parameters:
*  --disable-frame-rate-limit --disable-gpu-vsync

PG is https://playground.babylonjs.com/#NJ35D8#1:
* a single sphere / single material to minimize javascript time
* 16M faces for the sphere, material being setup to use all the PBR components (anisotropy, sheen, clearcoat, etc) so that "function calls" are maximized

Result: exact same fps with and without code refactoring.

It is 50fps (with the default playground window size when the browser is maximized) for 10/15s then stabilizes to 45fps.

Sames results in Firefox.

**Comment by sebavan** [comment]
the thing is here you would be vertex bound ? Not necessarily fragment which is the part we are refactoring?

**Comment by Popov72** [comment]
Yes indeed. Going to draw a single plane multiple times and see how it comes.

**Comment by Popov72** [comment]
PG: https://playground.babylonjs.com/#V3A42Y

Only 102 tris drawn per frame.

It is running at 23 fps in both cases, with and without code refactoring.

Same one but without the direct lighting code and with sheen/clear coat/sub surface disabled (those are the biggest blocks, with quite a lot of code in), to maximize the number of "function calls" vs total code size: https://playground.babylonjs.com/#V3A42Y#1

Running at 35fps in both cases.

**Comment by sebavan** [comment]
Sooooo coooooool, could you try on a mobile just to be fully safe ? Sometimes their optim step is not as good :-)

**Comment by Popov72** [comment]
I don't know how it access a local url on my mobile, as the refactored code is only available there...

**Comment by sebavan** [comment]
I think you can npm start --public to make your local dev env available accross your local network.

If it does not work, I ll check locally to see the results :-)

**Comment by Popov72** [comment]
It seems `--public` does not work, I have only TCP listeners on the 1338 port for localhost (127.0.0.1):

```
C:\>netstat -a

Connexions actives

  Proto  Adresse locale         Adresse distante       État
  TCP    0.0.0.0:135            DESKTOP-J4K5S30:0      LISTENING
  TCP    0.0.0.0:445            DESKTOP-J4K5S30:0      LISTENING
  TCP    0.0.0.0:5040           DESKTOP-J4K5S30:0      LISTENING
  TCP    0.0.0.0:5357           DESKTOP-J4K5S30:0      LISTENING
  TCP    0.0.0.0:27036          DESKTOP-J4K5S30:0      LISTENING
  TCP    0.0.0.0:49664          DESKTOP-J4K5S30:0      LISTENING
  TCP    0.0.0.0:49665          DESKTOP-J4K5S30:0      LISTENING
  TCP    0.0.0.0:49666          DESKTOP-J4K5S30:0      LISTENING
  TCP    0.0.0.0:49667          DESKTOP-J4K5S30:0      LISTENING
  TCP    0.0.0.0:49668          DESKTOP-J4K5S30:0      LISTENING
  TCP    0.0.0.0:49669          DESKTOP-J4K5S30:0      LISTENING
  TCP    0.0.0.0:49672          DESKTOP-J4K5S30:0      LISTENING
  TCP    0.0.0.0:49673          DESKTOP-J4K5S30:0      LISTENING
  TCP    127.0.0.1:1338         DESKTOP-J4K5S30:0      LISTENING
  TCP    127.0.0.1:5354         DESKTOP-J4K5S30:0      LISTENING
  TCP    127.0.0.1:5354         DESKTOP-J4K5S30:49670  ESTABLISHED
  TCP    127.0.0.1:5354         DESKTOP-J4K5S30:49671  ESTABLISHED
  TCP    127.0.0.1:5354         DESKTOP-J4K5S30:49792  ESTABLISHED
  TCP    127.0.0.1:5354         DESKTOP-J4K5S30:49793  ESTABLISHED
  TCP    127.0.0.1:5354         DESKTOP-J4K5S30:53639  ESTABLISHED
  TCP    127.0.0.1:27015        DESKTOP-J4K5S30:0      LISTENING
  TCP    127.0.0.1:27015        DESKTOP-J4K5S30:49731  ESTABLISHED
  TCP    127.0.0.1:27015        DESKTOP-J4K5S30:49745  ESTABLISHED
  TCP    127.0.0.1:27015        DESKTOP-J4K5S30:49840  ESTABLISHED
  TCP    127.0.0.1:27015        DESKTOP-J4K5S30:53627  ESTABLISHED
  TCP    127.0.0.1:27015        DESKTOP-J4K5S30:53646  ESTABLISHED
  TCP    127.0.0.1:27015        DESKTOP-J4K5S30:53705  ESTABLISHED
  TCP    127.0.0.1:27060        DESKTOP-J4K5S30:0      LISTENING
  TCP    127.0.0.1:49442        DESKTOP-J4K5S30:65001  ESTABLISHED
  TCP    127.0.0.1:49670        DESKTOP-J4K5S30:5354   ESTABLISHED
  TCP    127.0.0.1:49671        DESKTOP-J4K5S30:5354   ESTABLISHED
  TCP    127.0.0.1:49723        DESKTOP-J4K5S30:0      LISTENING
  TCP    127.0.0.1:49731        DESKTOP-J4K5S30:27015  ESTABLISHED
  TCP    127.0.0.1:49745        DESKTOP-J4K5S30:27015  ESTABLISHED
  TCP    127.0.0.1:49792        DESKTOP-J4K5S30:5354   ESTABLISHED
  TCP    127.0.0.1:49793        DESKTOP-J4K5S30:5354   ESTABLISHED
  TCP    127.0.0.1:49840        DESKTOP-J4K5S30:27015  ESTABLISHED
  TCP    127.0.0.1:50647        DESKTOP-J4K5S30:50648  ESTABLISHED
  TCP    127.0.0.1:50648        DESKTOP-J4K5S30:50647  ESTABLISHED
  TCP    127.0.0.1:52193        DESKTOP-J4K5S30:0      LISTENING
  TCP    127.0.0.1:53627        DESKTOP-J4K5S30:27015  ESTABLISHED
  TCP    127.0.0.1:53639        DESKTOP-J4K5S30:5354   ESTABLISHED
  TCP    127.0.0.1:53646        DESKTOP-J4K5S30:27015  ESTABLISHED
  TCP    127.0.0.1:53705        DESKTOP-J4K5S30:27015  ESTABLISHED
  TCP    127.0.0.1:59243        DESKTOP-J4K5S30:0      LISTENING
  TCP    127.0.0.1:65000        DESKTOP-J4K5S30:0      LISTENING
  TCP    127.0.0.1:65001        DESKTOP-J4K5S30:0      LISTENING
  TCP    127.0.0.1:65001        DESKTOP-J4K5S30:49442  ESTABLISHED
  TCP    172.20.10.2:139        DESKTOP-J4K5S30:0      LISTENING
```
172.20.10.2 is my IP (my iPhone is connected to the PC and provides the internet connection).

**Comment by Popov72** [comment]
Ok, managed to make it work with:
`npm run start-public-ssl`

When you browse the page you get a security warning but you can still proceed.

So going for the tests now.

**Comment by Popov72** [comment]
Results with a iPhone SE:
* https://playground.babylonjs.com/#3USSLS#2
  * 30/31 fps without code refactoring, 34/35 fps with refactoring!
* https://playground.babylonjs.com/#NF0X6N#1
  * 42 fps without code refactoring, 41 fps with refactoring

**Comment by sebavan** [comment]
So all good for me then !!!

**Pull request merged**